File: lexicopraphicalnumbers.py
class Solution(object):
    def lexicalOrder(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        res = []
        for i in xrange(1,10):
            self.dfs(i, n, res)
        return res
    def dfs(self, cur ,n, res):
        if cur > n:
            return
        res.append(cur)
        for i in xrange(10):
            if cur*10+i > n:
                return
            self.dfs(cur*10+i, n, res)

File: cpp/lexicopraphicalnumbers.cpp
class Solution {
public:
    vector<int> lexicalOrder(int n) {
        vector<int> res;
        for (int i = 1; i < 10; i++) {
            if (i<=n) {
                helper(res, i, n);
            }
        }
        return res;
        
    }
    
    void helper (vector<int> &res, int i, int n) {
        res.push_back(i);
        for (int j = 0; j < 10; j++){
            if (i*10+j <= n) helper(res, i*10+j, n);
            else return;
        }
    }
    
};
